wiki.moinmoin

```diff
@@ -385,15 +384,58 @@ RULE:     - the owning team will provide timely, high quality updates for the
+RULE:       a) The need to adapt the current version of the main package and/or
+RULE:          other vendored dependencies to work with the new dependency
+RULE:       b) The need to backport the fix in the dependency as the main
+RULE:          package will functionally only work well with the older version
+RULE:       c) The need to backport the fix in the dependency, as it would imply
+RULE:          requiring a newer toolchain to be buildable that isn't available
+RULE:          in the target release.
+RULE: - The rust ecosystem currently isn't yet considered stable enough for
+RULE:   classic lib dependencies and transitions in main; therefore the
+RULE:   expectation for those packages is to vendor (and own/test) all
+RULE:   dependencies (except those provided by the rust runtime itself).
+RULE:   This implies that all the rules for vendored builds always
+RULE:   apply to them. In addition:
+RULE:   - The rules and checks for rust based packges are preliminary and might
+RULE:     change while finalizing the rust toolchain in main and while
+RULE:     processing the first few rust based packages.
+RULE:   - Rust builds using librust-*-dev packges will populate the attribute
+RULE:     `Built-Using`, right now this is expected not to be present as
+RULE:      - for now - all rust dependencies shall be vendored.
+RULE:   - It is expected rust builds will use dh-cargo so that a later switch
+RULE:     to non vendored dependencies isn't too complex (e.g. it is likely
+RULE:     that over time more common libs shall become stable and then archive
+RULE:     packages will be used to build).
+RULE: - All vendored dependencies (no matter what language) shall have a
+RULE:   way to be refreshed
 TODO-A: - This does not use static builds
 TODO-B: - The team <TBD> is aware of the implications by a static build and
 TODO-B:   commits to test no-change-rebuilds and to fix any issues found for the
 TODO-B:   lifetime of the release (including ESM)
+
 TODO-A: - This does not use vendored code
 TODO-B: - The team <TBD> is aware of the implications of vendored code and (as
-TODO-B:   alerted by the security team) commits to provide updates to the security
-TODO-B:   team for any affected vendored code for the lifetime of the release
-TODO-B:   (including ESM).
+TODO-B:   alerted by the security team) commits to provide updates and backports
+TODO-B:   to the security team for any affected vendored code for the lifetime
+TODO-B:   of the release (including ESM).
+
+TODO-A: - This does not use vendored code
+TODO-B: - This package uses vendored go code tracked in go.mod as shiped in the
+TODO-B:   package, refreshing that code works via `go mod ...`.
+TODO-C: - This package uses vendored rust code tracked in Cargo.lock as shipped,
+TODO-C:   in the package, refreshing that code works via `cargo update ...`.
+TODO-D: - This package uses vendored code, refreshing that code is outlined
+TODO-D:   in debian/README.source
```

Comment:
Based on feedback in the rustc MIR, I suggest including a requirement for the package to list all its vendored sources in a field `XS-Vendored-Sources-$language`, formatted as a comma delimited list of `$package@$version` items.
The reason for having this field language-specific is that each language has its own package namespace. For instance, you can find a "ipaddr" package in a *lot* of languages.

However, I'm yet to find a more satisfying mechanism than having the maintainer manually fill the fields themselves, alongside a sanity check in d/rules. Both the generation script and the check can of course live in dh-cargo for Rust.

Ideally we'd use debian/substvars for this, but apparently it's not taken into account by dpkg-source by default, which makes it even more awkward as the caller now has to figure out how to pass the correct options to whatever they're using for their package.

@setharnold I'd love your feedback on this?

Reply:
I'm not very familiar with the dh_ family of tools :(

I think anything that relies upon maintainer action / initiative is going to be doomed to be out of date. It really does need to be "free" for the packager; this seems like a problem the Debian golang and rust teams are interested in solving. I wonder if there's a good way for dh to provide most of the infrastructure for it, and try to simplify the language-specific portions?

This might be way off base though, as I said, I'm vastly unfamiliar with dh's internals.

Reply:
I think I need to expand a little bit on some details that I glossed over.

First, we're talking about tracking *vendored* dependencies, which is expected to be a requirement for Rust package in `main`, and is already the case for Go package AIUI. The Debian discussion is about other source packages that are statically linked into the binary, which is NOT the same thing.

Second, this is a source-level property of the package, and the generated data can be quite massive. This is why we cant it in the Sources index, which isn't downloaded by our users by default, hence the use of a XS field, which in turns led to my manual update predicament, as this means it cannot be part of the build process.

However, the whole point of having a sanity check in the debian/rules is to avoid the data going stale. Thinking about it some more, the symbols files work the same way, with the packager having to do a manual update but is guided by the tooling. The idea is indeed to leverage the packaging tools (i.e. dh-cargo for Rust) to automate this. Of course, the first implementation of the field being rustc, which has custom packaging, I'll do it explicitly in the debian/rules there. I still need to start working on vendoring support altogether in dh-cargo anyway ;-).

Reply:
I'd love to also have vendoring support in dh-golang. The Debian tooling for golang predates golang's modules design and therefore predates a formal notion of vendoring. Due to this we end up packaging the source code for tons of 'vendored' requirements in separate packages. As an example we have golang-github-shirou-gopsutil. It is a package that ships no binaries, only source code. This is used to build another package, syncthing, that builds actual binaries. Changing this over to a dh-golang supported vendoring design would be great.

Reply:
To be fair, regardless of golang formal notion of vendoring, the Debian tooling would have sought to split vendored dependencies in separate packages, as that's deemed the most maintainable way from a security PoV (patch once, rebuild everything). However, we revert this for at least some of our packages, e.g. `runc` which has a ton of golang deps in Debian but instead vendors them in Ubuntu.

What I'm proposing here is just formalizing how we track those vendored dependencies so that we know what needs patching, and once there's agreement on the format adding support for it in the languages tooling.